File: analysis_gpt.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
#include <map>
#include <chrono>
#include <iomanip>

using namespace std;
using namespace std::chrono;

// Tính tỉ lệ diện tích của bounding box so với frame (theo %)
double calculateAreaRatio(const vector<float>& box, int frameWidth, int frameHeight) {
    double boxWidth = box[2] - box[0];
    double boxHeight = box[3] - box[1];
    double boxArea = boxWidth * boxHeight;
    double frameArea = frameWidth * frameHeight;
    return (boxArea / frameArea) * 100.0;
}

// Tính trung bình (mean)
double calculateMean(const vector<double>& ratios) {
    if (ratios.empty()) return 0.0;
    double sum = 0.0;
    for (double val : ratios) sum += val;
    return sum / ratios.size();
}

// Tính trung vị (median)
double calculateMedian(vector<double> ratios) {
    if (ratios.empty()) return 0.0;
    sort(ratios.begin(), ratios.end());
    size_t n = ratios.size();
    if (n % 2 == 0)
        return (ratios[n / 2 - 1] + ratios[n / 2]) / 2.0;
    else
        return ratios[n / 2];
}

// Tính mode (phổ biến nhất) với sai số (tolerance)
double calculateMode(const vector<double>& ratios, double tolerance = 1.0) {
    if (ratios.empty()) return 0.0;
    map<int, int> freq;
    for (double val : ratios) {
        int rounded = static_cast<int>(round(val / tolerance));
        freq[rounded]++;
    }

    int maxCount = 0;
    int modeRounded = 0;
    for (const auto& [key, count] : freq) {
        if (count > maxCount) {
            maxCount = count;
            modeRounded = key;
        }
    }

    return modeRounded * tolerance;
}

// Phân tích arrival/departure dựa trên các ngưỡng
string detectArrivalOrDeparture(
    int humanCount,
    int nonHumanCount,
    bool currentHumanDetected,
    double medianRatio,
    double modeRatio,
    double arrivalMedianThreshold,
    double arrivalModeThreshold,
    double departureMedianThreshold,
    double departureModeThreshold,
    int arrivalFrameThreshold,
    int departureFrameThreshold
) {
    if (humanCount >= arrivalFrameThreshold &&
        !currentHumanDetected &&
        medianRatio > arrivalMedianThreshold &&
        modeRatio > arrivalModeThreshold) {
        return "arrival";
    }

    if (nonHumanCount > departureFrameThreshold &&
        currentHumanDetected &&
        medianRatio < departureMedianThreshold &&
        modeRatio < departureModeThreshold) {
        return "departure";
    }

    return "none";
}

// // Test ví dụ
// int main() {
//     vector<vector<float>> boxes = {
//         {100, 100, 300, 300},
//         {150, 120, 320, 310},
//         {130, 110, 310, 305}
//     };

//     int frameWidth = 640;
//     int frameHeight = 480;
//     vector<double> areaRatios;

//     for (const auto& box : boxes) {
//         double ratio = calculateAreaRatio(box, frameWidth, frameHeight);
//         areaRatios.push_back(ratio);
//     }

//     double meanRatio = calculateMean(areaRatios);
//     double medianRatio = calculateMedian(areaRatios);
//     double modeRatio = calculateMode(areaRatios, 1.0);

//     cout << fixed << setprecision(2);
//     cout << "Mean area ratio: " << meanRatio << "%" << endl;
//     cout << "Median area ratio: " << medianRatio << "%" << endl;
//     cout << "Mode area ratio: " << modeRatio << "%" << endl;

//     string result = detectArrivalOrDeparture(
//         25, // human count
//         5,  // non-human count
//         false, // current human detected
//         medianRatio,
//         modeRatio,
//         20.0, // arrival median threshold
//         15.0, // arrival mode threshold
//         10.0, // departure median threshold
//         8.0,  // departure mode threshold
//         20,   // arrival frame threshold
//         18    // departure frame threshold
//     );

//     cout << "Detected event: " << result << endl;

//     return 0;
// }

File: analysis_gpt.py
import statistics
from typing import List, Tuple, Optional
from collections import deque
from datetime import datetime

# --- Hàm tính tỷ lệ diện tích box so với frame ---
def calculate_area_ratio(box: List[float], frame_shape: Tuple[int, int, int]) -> float:
    frame_area = frame_shape[0] * frame_shape[1]  # height * width
    box_area = (box[2] - box[0]) * (box[3] - box[1])  # (xmax - xmin) * (ymax - ymin)
    return (box_area / frame_area) * 100.0  # trả về phần trăm

# --- Hàm tính thống kê ---
def calculate_statistics(ratios: List[float], mode_tolerance: float = 1.0) -> Tuple[float, float, float]:
    if not ratios:
        return 0.0, 0.0, 0.0

    mean_ratio = sum(ratios) / len(ratios)
    median_ratio = statistics.median(ratios)

    # Làm tròn theo tolerance để tìm mode
    rounded_ratios = [round(r / mode_tolerance) * mode_tolerance for r in ratios]
    try:
        mode_ratio = statistics.mode(rounded_ratios)
    except statistics.StatisticsError:
        mode_ratio = median_ratio  # fallback nếu không có mode rõ ràng

    return mean_ratio, median_ratio, mode_ratio

# --- Hàm xác định người đến hoặc rời ---
def detect_arrival_or_departure(
    detection_ratios: deque,
    frame_time: datetime,
    human_count: int,
    non_human_count: int,
    current_human_detected: bool,
    config: dict
) -> Optional[str]:
    """
    Trả về: 'arrival', 'departure', hoặc None
    """

    mean_ratio, median_ratio, mode_ratio = calculate_statistics(list(detection_ratios))

    # Điều kiện người đến
    if (human_count >= config['welcome_threshold'] and
        not current_human_detected and
        median_ratio > config['median_ratio_threshold'] and
        mode_ratio > config['mode_ratio_threshold']):
        print(f"[{frame_time}] 👤 Người đến - Median: {median_ratio:.2f}% - Mode: {mode_ratio:.2f}%")
        return 'arrival'

    # Điều kiện người rời
    elif (non_human_count >= config['departure_threshold'] and
          current_human_detected and
          median_ratio < config['departure_median_threshold'] and
          mode_ratio < config['departure_mode_threshold']):
        print(f"[{frame_time}] 👋 Người rời - Median: {median_ratio:.2f}% - Mode: {mode_ratio:.2f}%")
        return 'departure'

    return None
